pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,309 @@
+	"MEI_DEV_ID_SPT_2":      0x9D3B,
+	"MEI_DEV_ID_SPT_H":      0xA13A,
+	"MEI_DEV_ID_SPT_H_2":    0xA13B,
+	"MEI_DEV_ID_LBG":        0xA1BA,
+	"MEI_DEV_ID_BXT_M":      0x1A9A,
+	"MEI_DEV_ID_APL_I":      0x5A9A,
+	"MEI_DEV_ID_DNV_IE":     0x19E5,
+	"MEI_DEV_ID_GLK":        0x319A,
+	"MEI_DEV_ID_KBP":        0xA2BA,
+	"MEI_DEV_ID_KBP_2":      0xA2BB,
+	"MEI_DEV_ID_CNP_LP":     0x9DE0,
+	"MEI_DEV_ID_CNP_LP_4":   0x9DE4,
+	"MEI_DEV_ID_CNP_H":      0xA360,
+	"MEI_DEV_ID_CNP_H_4":    0xA364,
+	"MEI_DEV_ID_CMP_LP":     0x02e0,
+	"MEI_DEV_ID_CMP_LP_3":   0x02e4,
+	"MEI_DEV_ID_CMP_V":      0xA3BA,
+	"MEI_DEV_ID_CMP_H":      0x06e0,
+	"MEI_DEV_ID_CMP_H_3":    0x06e4,
+	"MEI_DEV_ID_CDF":        0x18D3,
+	"MEI_DEV_ID_ICP_LP":     0x34E0,
+	"MEI_DEV_ID_JSP_N":      0x4DE0,
+	"MEI_DEV_ID_TGP_LP":     0xA0E0,
+	"MEI_DEV_ID_MCC":        0x4B70,
+	"MEI_DEV_ID_MCC_4":      0x4B75,
+}
+
+// various ME-related constants
+const (
+	// ME - current working state set to normal
+	MEHfs1CwsNormal = 0x5
+	// ME - current operation mode set to normal
+	MEHfs1ComNormal = 0x0
+	// ME - CSE's firmware SKU is custom
+	MEHfs3FwSkuCustom = 0x5
+)
+
+// MKHIClient is a client to send MKHI commands via MEI.
+type MKHIClient struct {
+	MEI              MEI
+	clientProperties *ClientProperties
+}
+
+// MKHI command groups, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L22
+const (
+	MKHIGroupIDCbm    = 0x0
+	MKHIGroupIDHmrfpo = 0x5
+	MKHIGroupIDGen    = 0xff
+)
+
+// MKHI HMRFPO command IDs, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L33
+const (
+	MKHIHmrfpoEnable    = 0x1
+	MKHIHmrfpoGetStatus = 0x3
+)
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
+	parts := strings.Split(pciaddr, ":")
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("wrong PCI address string '%s': want 3 colon-separated groups, got %d", pciaddr, len(parts))
+	}
+	parts = strings.Split(parts[2], ".")
+	if len(parts) != 2 {
+		return 0, fmt.Errorf("wrong PCI address sub-component '%s': want 2 dot-separated groups, got %d", pciaddr, len(parts))
+	}
+	a, err := strconv.ParseInt(parts[0], 16, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse hex string '%s': %v", parts[0], err)
+	}
+	return int(a), nil
+}
+
+// Open opens an MKHI client connection.
+func (m *MKHIClient) Open(meiPath string) error {
+	cp, err := m.MEI.Open(meiPath, MKHIGuid)
+	if err != nil {
+		return err
+	}
+	m.clientProperties = cp
+	return nil
+}
+
+// Close closes an MKHI client connection.
+func (m *MKHIClient) Close() error {
+	m.clientProperties = nil
+	return m.MEI.Close()
+}
+
+// IsHmrfpoEnableAllowed queries whether the HMRFPO enable is allowed.
+func (m MKHIClient) IsHmrfpoEnableAllowed() (bool, error) {
```

Comment:
give the client a pointer, *MKHIClient, or you'll be passing lots of data around with a call to this